# Machine Learning Environment for the Isabelle Proof Assistant

<div style="max-width: 1000px;">

## Demostration of usage
To demostrate the ability of small-step and big-step proof correctness verification in training and evaluating of Prover, we select five questions from miniF2F, raw data formated in the form of LEGO-prover data.
### Questions
|problem name | informal statement | formal statement |
| ----------- | ------------------ | ---------------- |
| mathd_numbertheory_3 | What is the units digit of the sum of the squares of the first nine positive integers? Show that it is 5. | ```theorem mathd_numbertheory_3 :\n  \"(\<Sum> x < 10. ((x + 1)^2)) mod 10 = (5::nat)\"``` |
| mathd_algebra_419 | What is the value of $-a-b^2+3ab$ if $a=-1$ and $b=5$? Show that it is -39. | ```theorem mathd_algebra_419:\n  fixes a b :: real\n  assumes \"a = -1\"\n and \"b = 5\"\n  shows \"-a - b^2 + 3 * (a * b) = -39\"``` |
| numbertheory_3pow2pownm1mod2pownp3eq2pownp2 | Show that for positive integer $n$, the remainder of $3^{2^n} - 1 \equiv 2^{n+2} \mod 2^{n+3}$. | ```theorem numbertheory_3pow2pownm1mod2pownp3eq2pownp2:\n  fixes n :: nat\n  assumes \"0 < n\" \n  shows \"(3^(2^n) - 1) mod (2^(n + 3)) = (2::nat)^(n + 2)\"``` |
| aime_1983_p1 | Let $x$, $y$ and $z$ all exceed $1$ and let $w$ be a positive number such that $\\log_x w = 24$, $\\log_y w = 40$ and $\\log_{xyz} w = 12$. Find $\\log_z w$. Show that it is 060. | ```theorem aime_1983_p1:\n  fixes x y z w :: nat\n  assumes ht : \"1 < x \\<and> 1 < y \\<and> 1 < z\"\n    and hw : \"0 \\<le> w\"\n    and h0 : \"ln w / ln x = 24\"\n    and h1 : \"ln w / ln y = 40\"\n    and h2 : \"ln w / ln (x * y * z) = 12\"\n  shows \"ln w / ln z = 60\"``` |
| amc12b_2020_p22 | What is the maximum value of $\frac{(2^t-3t)t}{4^t}$ for real values of $t?$ $\textbf{(A)} \frac{1}{16} \qquad\textbf{(B)} \frac{1}{15} \qquad\textbf{(C)} \frac{1}{12} \qquad\textbf{(D)} \frac{1}{10} \qquad\textbf{(E)} \frac{1}{9}$ Show that it is $\textbf{(C)} \frac{1}{12}$. | ```theorem amc12b_2020_p22:\n  fixes t :: real\n  shows \"((2 powr t - 3 * t) * t) / (4 powr t) \<le> 1 / 12\"``` |
</div>

In [1]:
# Raw data
questions = [
    {
        "problem_name": "mathd_numbertheory_3",
        "category": "unknown",
        "metadata": {},
        "informal_statement": "What is the units digit of the sum of the squares of the first nine positive integers? Show that it is 5.",
        "informal_proof": "We can reduce the amount of work we have to do in this problem by realizing that the units digit of the sum of the squares is the units digit of the sum of the units digits of the squares. In other words, the units digit of $1^2+2^2+\\ldots+9^2$ is the units digit of $1+4+9+6+5+6+9+4+1=45$, which is $5$.",
        "formal_statement": "theorem mathd_numbertheory_3 :\n  \"(\\<Sum> x < 10. ((x + 1)^2)) mod 10 = (5::nat)\"",
        "formal_code": "(*\n  Authors: Wenda Li\n*)\n\ntheory mathd_numbertheory_3 imports Complex_Main\nbegin\n\ntheorem mathd_numbertheory_3 :\n  \"(\\<Sum> x \\<le> 9. (x^2)) mod 10 = (5::nat)\"\n  by eval\n\nend "
    },
    {
        "problem_name": "mathd_algebra_419",
        "category": "unknown",
        "metadata": {},
        "informal_statement": "What is the value of $-a-b^2+3ab$ if $a=-1$ and $b=5$? Show that it is -39.",
        "informal_proof": "Plugging in the given values yields $-a-b^2+3ab=-(-1)-5^2+3(-1)(5)=1-25-15=-39$.",
        "formal_statement": "theorem mathd_algebra_419:\n  fixes a b :: real\n  assumes \"a = -1\"\n    and \"b = 5\"\n  shows \"-a - b^2 + 3 * (a * b) = -39\"",
        "formal_code": "(*\n  Authors: Wenda Li\n*)\n\ntheory mathd_algebra_419 imports Complex_Main\nbegin\n\ntheorem mathd_algebra_419:\n  fixes a b :: real\n  assumes \"a = -1\"\n    and \"b = 5\"\n  shows \"-a - b^2 + 3 * (a * b) = -39\"\n  using assms by auto\n\nend"
    },
    {
        "problem_name": "numbertheory_3pow2pownm1mod2pownp3eq2pownp2",
        "category": "unknown",
        "metadata": {},
        "informal_statement": "Show that for positive integer $n$, the remainder of $3^{2^n} - 1 \\equiv 2^{n+2} \\mod 2^{n+3}$.",
        "informal_proof": "We can prove this by induction on n. For the base case, $n=1$, when the result obviously holds.\nFor the inductive case, the hypothesis is $3^{2^n} - 1 \\equiv 2^{n+2} \\mod 2^{n+3}$, so we can obtain a natural number $p$ where $3^{2^n} - 1 = 2^{n+2} + p * 2^{n+3} = 2^{n+2}(1+2p)$.\nThen we have $3^{2^{n+1}} = 3^{2^n*2} = (3^{2^n})^2 = (2^{n+2} * (1+2p) + 1)^2$.\nHence $3^{2^{n+1}} \\equiv (2^{n+2} * (1+2p))^2 + 2 * 2^{n+2} * (1+2p) + 1 \\equiv 2^{n+3}*(1+2p) + 1 \\equiv 1\\mod 2^{n+3}$.\nTherefore $3^{2^{n+1}} -1 \\equiv 0 \\mod 2^{n+3}$. By induction, the statement holds true.",
        "formal_statement": "theorem numbertheory_3pow2pownm1mod2pownp3eq2pownp2:\n  fixes n :: nat\n  assumes \"0 < n\" \n  shows \"(3^(2^n) - 1) mod (2^(n + 3)) = (2::nat)^(n + 2)\"",
        "formal_code": "(*\n  Authors: Wenda Li\n*)\n\ntheory numbertheory_3pow2pownm1mod2pownp3eq2pownp2 \n  imports Complex_Main \"HOL-Computational_Algebra.Computational_Algebra\"\n  \"HOL-Number_Theory.Number_Theory\"\nbegin\n\ntheorem numbertheory_3pow2pownm1mod2pownp3eq2pownp2:\n  fixes n :: nat\n  assumes \"0 < n\" \n  shows \"(3^(2^n) - 1) mod (2^(n + 3)) = (2::nat)^(n + 2)\"\n  using assms\nproof (induct n)\n  case (Suc n)\n  have ?case when \"n=0\" \n    using that by simp\n  moreover have ?case when \"n>0\"\n  proof -\n    define m::nat where \"m = 2^n\"\n    have \"(3 ^ 2 ^ n - 1) mod 2 ^ (n + 3) = (2::nat)^ (n + 2)\"\n      using Suc(1) that by simp\n    then have \"(3^m - 1) mod (8*m) = 4*m\"\n      unfolding m_def by (auto simp:algebra_simps power_add) \n    then obtain k where k0:\"3^m - 1 = (8*m)*k + 4*m\" \n      by (metis mult_div_mod_eq)\n    define M::nat where \"M = 4*m*k^2+m+4*m*k+k\"\n\n    have k:\"3^m = (8*m)*k + 4*m +1\" \n    proof -\n      have \"3^m\\<ge>(1::nat)\" by auto\n      then show ?thesis\n        using k0 by linarith\n    qed\n    have \"3 ^ 2 ^ Suc n - 1 = ((3::nat) ^ (m*2)) - 1\"\n      unfolding m_def by (auto simp:algebra_simps)\n    also have \"... = (3 ^ m)\\<^sup>2 - 1\"\n      unfolding power_mult by simp\n    also have \"... = ((8*m)*k + 4*m +1)^2 -1 \"\n      unfolding k by simp\n    also have \"... = (16*m)*M + 8*m\"\n      unfolding M_def by (auto simp:algebra_simps power2_eq_square)\n    finally have \"3 ^ 2 ^ Suc n - 1 = (16*m)*M + 8*m\" .\n    moreover have \"((16*m)*M + 8*m) mod (16*m) = 8*m\"\n      by simp\n    then have \"((16*m)*M + 8*m) mod 2 ^ (Suc n + 3) = 2 ^ (Suc n + 2)\"\n      unfolding m_def by (auto simp:algebra_simps power_add)\n    ultimately show ?thesis by auto\n  qed\n  ultimately show ?case by blast\nqed simp\n\nend   "
    },
    {
        "problem_name": "aime_1983_p1",
        "category": "unknown",
        "metadata": {},
        "informal_statement": "Let $x$, $y$ and $z$ all exceed $1$ and let $w$ be a positive number such that $\\log_x w = 24$, $\\log_y w = 40$ and $\\log_{xyz} w = 12$. Find $\\log_z w$. Show that it is 060.",
        "informal_proof": "The [[logarithm]]ic notation doesn't tell us much, so we'll first convert everything to the equivalent exponential forms.\n\n$x^{24}=w$, $y^{40}=w$, and $(xyz)^{12}=w$. If we now convert everything to a power of $120$, it will be easy to isolate $z$ and $w$.\n\n$x^{120}=w^5$, $y^{120}=w^3$, and $(xyz)^{120}=w^{10}$.\n\nWith some substitution, we get $w^5w^3z^{120}=w^{10}$ and $\\log_zw=060$.",
        "formal_statement": "theorem aime_1983_p1:\n  fixes x y z w :: nat\n  assumes ht : \"1 < x \\<and> 1 < y \\<and> 1 < z\"\n    and hw : \"0 \\<le> w\"\n    and h0 : \"ln w / ln x = 24\"\n    and h1 : \"ln w / ln y = 40\"\n    and h2 : \"ln w / ln (x * y * z) = 12\"\n  shows \"ln w / ln z = 60\"",
        "formal_code": "(*\n  Authors: Wenda Li\n*)\n\ntheory aime_1983_p1 imports Complex_Main\nbegin\n\ntheorem aime_1983_p1:\n  fixes x y z w :: nat\n  assumes ht : \"1 < x \\<and> 1 < y \\<and> 1 < z\"\n    and hw : \"0 \\<le> w\"\n    and h0 : \"ln w / ln x = 24\"\n    and h1 : \"ln w / ln y = 40\"\n    and h2 : \"ln w / ln (x * y * z) = 12\"\n  shows \"ln w / ln z = 60\"\nproof -\n  define xx yy zz ww where \"xx=ln x\" and \"yy = ln y\" \n      and \"zz = ln z\" and \"ww = ln w\"\n  have \"xx= ww/24\" using h0 ht unfolding xx_def ww_def \n    by (auto simp:field_simps)\n  moreover have \"yy = ww / 40\" using ht h1 unfolding yy_def ww_def\n    by (auto simp:field_simps)\n  moreover have \"xx+yy+zz > 0\" \n    unfolding xx_def yy_def zz_def using ht \n    by (metis ln_gt_zero of_nat_1 of_nat_less_iff pos_add_strict)\n  then have \"ww = 12*(xx+yy+zz)\"using ht h2\n    unfolding xx_def yy_def zz_def ww_def\n    by (auto simp:field_simps ln_mult)\n  ultimately have \"ww = 12*(ww/24 + ww/40 + zz)\"\n    by blast\n  then have \"ww=60*zz\"\n    by (auto simp:field_simps)\n  then show ?thesis unfolding ww_def zz_def using ht by auto\nqed\n\nend\n"
    },
    {
        "problem_name": "amc12b_2020_p22",
        "category": "unknown",
        "metadata": {},
        "informal_statement": "What is the maximum value of $\\frac{(2^t-3t)t}{4^t}$ for real values of $t?$\n\n$\\textbf{(A)}\\ \\frac{1}{16} \\qquad\\textbf{(B)}\\ \\frac{1}{15} \\qquad\\textbf{(C)}\\ \\frac{1}{12} \\qquad\\textbf{(D)}\\ \\frac{1}{10} \\qquad\\textbf{(E)}\\ \\frac{1}{9}$ Show that it is \\textbf{(C)} \\frac{1}{12}.",
        "informal_proof": "We proceed by using AM-GM. We get $\\frac{(2^t-3t) + 3t}{2}$ $\\ge \\sqrt{(2^t-3t)(3t)}$. Thus, squaring gives us that $4^{t-1} \\ge (2^t-3t)(3t)$. Rembering what we want to find, we divide both sides of the inequality by the positive amount of $\\frac{1}{3\\cdot4^t}$. We get the maximal values as $\\frac{1}{12}$, and we are done.",
        "formal_statement": "theorem amc12b_2020_p22:\n  fixes t :: real\n  shows \"((2 powr t - 3 * t) * t) / (4 powr t) \\<le> 1 / 12\"",
        "formal_code": "theory amc12b_2020_p22\n  imports Complex_Main \"HOL-Library.Code_Target_Numeral\"\nbegin\n\ntheorem amc12b_2020_p22: fixes t :: real shows \"((2 powr t - 3 * t) * t) / (4 powr t) \<le> 1 / 12\" proof - have h1: \"4 * ((2 powr t - 3 * t) * (3 * t)) \<le> (2 powr t)^2\" by (smt (verit) power2_sum ring_class.ring_distribs(2) sum_squares_bound) have hpow: \"(2 powr t)^2 = 4 powr t\" by (metis powr_mult numeral_Bit0_eq_double power2_eq_square) have h2: \"12 * ((2 powr t - 3 * t) * t) \<le> 4 powr t\" using h1 hpow by simp have pos: \"0 < 12 * (4 powr t)\" by simp have eq: \"((2 powr t - 3 * t) * t) / (4 powr t) = (12 * ((2 powr t - 3 * t) * t)) / (12 * (4 powr t))\" by (simp add: field_simps) have \"(12 * ((2 powr t - 3 * t) * t)) / (12 * (4 powr t)) \<le> (4 powr t) / (12 * (4 powr t))\" using h2 pos by simp also have \"... = 1 / 12\" by simp finally show ?thesis using eq by simp qed end"
    }
]

In [2]:
from client.async_client import IsabelleGymAsyncClient
client = IsabelleGymAsyncClient()
proofs = [i.get("formal_code") for i in questions]
proofs

['(*\n  Authors: Wenda Li\n*)\n\ntheory mathd_numbertheory_3 imports Complex_Main\nbegin\n\ntheorem mathd_numbertheory_3 :\n  "(\\<Sum> x \\<le> 9. (x^2)) mod 10 = (5::nat)"\n  by eval\n\nend ',
 '(*\n  Authors: Wenda Li\n*)\n\ntheory mathd_algebra_419 imports Complex_Main\nbegin\n\ntheorem mathd_algebra_419:\n  fixes a b :: real\n  assumes "a = -1"\n    and "b = 5"\n  shows "-a - b^2 + 3 * (a * b) = -39"\n  using assms by auto\n\nend',
 '(*\n  Authors: Wenda Li\n*)\n\ntheory numbertheory_3pow2pownm1mod2pownp3eq2pownp2 \n  imports Complex_Main "HOL-Computational_Algebra.Computational_Algebra"\n  "HOL-Number_Theory.Number_Theory"\nbegin\n\ntheorem numbertheory_3pow2pownm1mod2pownp3eq2pownp2:\n  fixes n :: nat\n  assumes "0 < n" \n  shows "(3^(2^n) - 1) mod (2^(n + 3)) = (2::nat)^(n + 2)"\n  using assms\nproof (induct n)\n  case (Suc n)\n  have ?case when "n=0" \n    using that by simp\n  moreover have ?case when "n>0"\n  proof -\n    define m::nat where "m = 2^n"\n    have "(3 ^ 2 ^ n -

In [3]:
import sys
from pathlib import Path

ROOT = Path.cwd().parent
if str(ROOT) not in sys.path:
    sys.path.insert(0, str(ROOT))

In [6]:
from evaluation.scripts.theory_splitter import (
    extract_theory_name,
    split_commands,
)

## Bigstep verification

In [14]:
result = await client.verify_bigstep_text(theory_name="mathd_numbertheory_3", theory_text=proofs[0])
result

<Response [200 OK]>

In [ ]:
for proof in proofs:
    result = await client.verify_bigstep_text(theory_name=extract_theory_name(proof), theory_text=proof)
    print(result)

## Smallstep verification

### Single proof verification
For this demostration, we will need to split the proof to stepwise tactics to simulate the real use case in prover training and evaluation. This step is by the same splitter used in the evaluation.

In [23]:
session_payload = await client.acquire_session(["Complex_Main"], "HOL")
session_payload

{'session_id': '55232a7b-6a62-4a66-a546-6c6d8c27d69f',
 'created_at': 1779913230.0128596,
 'theories': ['Complex_Main'],
 'status': 'active',
 'reused': False,
 'lease_id': 'c8d82ff97d88'}

In [24]:
session_id = session_payload.get("session_id")
lease_id = session_payload.get("lease_id")
commands = split_commands(proofs[0])
commands

['(*\n  Authors: Wenda Li\n*)\n\ntheory mathd_numbertheory_3 imports Complex_Main\nbegin',
 'theorem mathd_numbertheory_3 :\n  "(\\<Sum> x \\<le> 9. (x^2)) mod 10 = (5::nat)"',
 'by eval',
 'end']

In [25]:
await client.enter_theory(session_id, "mathd_numbertheory_3", lease_id=lease_id)

{'success': True, 'message': 'Entered theory mathd_numbertheory_3'}

In [26]:
await client.execute_command(session_id, commands[0], lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': [],
 'execution_time': 0.11754894256591797,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [27]:
await client.execute_command(session_id, commands[1], lease_id=lease_id)

{'success': True,
 'output': 'val it = (): unit',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. sum power2 {..9} mod 10 = 5'],
 'execution_time': 0.1082608699798584,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [28]:
await client.execute_command(session_id, "sledgehammer", lease_id=lease_id)

{'success': True,
 'output': 'val it = (): unit\nval it = (): unit\nSledgehammering...\nNo proof found',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. sum power2 {..9} mod 10 = 5'],
 'execution_time': 11.7899489402771,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [25]:
await client.execute_command(session_id, commands[2], lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': [],
 'execution_time': 0.40366268157958984,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [26]:
await client.execute_command(session_id, commands[3], lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': [],
 'execution_time': 0.10254931449890137,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

## Test sledgehammer

Earlier we invoked sledgehammer by sending the literal `sledgehammer` Isar command as a proof **step** (`execute_command(..., 'sledgehammer')`). Running it that way executes inside the document node and blocks the session's other ML channels (e.g. `open_subgoals`) until it finishes.

Here we use the **standalone sledgehammer channel** instead: `client.sledgehammer(session_id, ...)` hits the dedicated `/sledgehammer` endpoint. On the ML side it forks an Isabelle thread, so it does **not** run as a proof step, does not advance/modify the proof state (read-only), and does not block other sessions. It returns the proof-method suggestions (e.g. `by (metis ...)`) which the prover can then apply as the next step.

In [14]:
session_payload = await client.acquire_session(["Complex_Main"], "HOL")
session_payload

{'session_id': '034b2d04-94ab-425d-9cbb-fb7ae93b76f7',
 'created_at': 1780463041.5500674,
 'theories': ['Complex_Main'],
 'status': 'active',
 'reused': True,
 'lease_id': '7690aba810aa'}

In [15]:
session_id = session_payload.get("session_id")
lease_id = session_payload.get("lease_id")

In [ ]:
# Test release session
await client.release_session(session_id, lease_id=lease_id)

{'success': True, 'session_id': '034b2d04-94ab-425d-9cbb-fb7ae93b76f7'}

In [16]:
await client.enter_theory(session_id, "amc12b_2020_p22", lease_id=lease_id)

{'success': True, 'message': 'Entered theory amc12b_2020_p22'}

In [17]:
await client.execute_command(session_id, 'theory amc12b_2020_p22 imports Complex_Main "HOL-Library.Code_Target_Numeral" begin', lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': [],
 'execution_time': 0.7222416400909424,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [18]:
await client.execute_command(session_id, 'theorem amc12b_2020_p22: fixes t :: real shows "((2 powr t - 3 * t) * t) / (4 powr t) ≤ 1 / 12"', lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. (2 powr t - 3 * t) * t / 4 powr t \\<le> 1 / 12'],
 'execution_time': 0.10210990905761719,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [19]:
await client.execute_command(session_id, 'proof -', lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. (2 powr t - 3 * t) * t / 4 powr t \\<le> 1 / 12'],
 'execution_time': 0.10212922096252441,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [20]:
await client.execute_command(session_id, 'have hpow: "(2 powr t)^2 = 4 powr t"', lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. (2 powr t)\\<^sup>2 = 4 powr t'],
 'execution_time': 0.10215520858764648,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [21]:
# Standalone sledgehammer channel: dedicated endpoint, not a proof step.
# Returns proof-method suggestions without modifying the proof state.
sh = await client.sledgehammer(session_id, timeout_s=60, lease_id=lease_id)
sh

{'success': True,
 'suggestions': ['metis found a proof...',
  'metis found a proof...',
  'metis: Try this: by (metis powr_mult numeral_Bit0_eq_double power2_eq_square) (7 ms)',
  'metis: Duplicate proof',
  'zipperposition found a proof...',
  'zipperposition: Try this: by (metis numeral_Bit0_eq_double power2_eq_square powr_mult) (31 ms)',
  'Done',
  'Try this: by (metis powr_mult numeral_Bit0_eq_double power2_eq_square) (7 ms)'],
 'raw_output': 'metis found a proof...\nmetis found a proof...\nmetis: Try this: by (metis powr_mult numeral_Bit0_eq_double power2_eq_square) (7 ms)\nmetis: Duplicate proof\nzipperposition found a proof...\nzipperposition: Try this: by (metis numeral_Bit0_eq_double power2_eq_square powr_mult) (31 ms)\nDone\nTry this: by (metis powr_mult numeral_Bit0_eq_double power2_eq_square) (7 ms)',
 'execution_time': 21.489171504974365}

In [22]:
await client.execute_command(session_id, 'by (metis powr_mult numeral_Bit0_eq_double power2_eq_square)', timeout=120.0, lease_id=lease_id)

{'success': True,
 'output': 'have hpow: (2 powr t)⇧2 = 4 powr t',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. (2 powr t - 3 * t) * t / 4 powr t \\<le> 1 / 12'],
 'execution_time': 0.11284089088439941,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [23]:
await client.execute_command(session_id, 'have pos: "(0::real) < 4 powr t" by simp', timeout=120.0, lease_id=lease_id)

{'success': True,
 'output': 'have pos: 0 < 4 powr t',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. (2 powr t - 3 * t) * t / 4 powr t \\<le> 1 / 12'],
 'execution_time': 0.10291004180908203,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [39]:
await client.execute_command(session_id, 'have amgm: "2 * (2 powr t) * (6 * t) ≤ (2 powr t)^2 + (6 * t)^2"', timeout=120.0, lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. 2 * 2 powr t * (6 * t) \\<le> (2 powr t)\\<^sup>2 + (6 * t)\\<^sup>2'],
 'execution_time': 0.10155200958251953,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [40]:
await client.execute_command(session_id, 'by (rule sum_squares_bound)', timeout=120.0, lease_id=lease_id)

{'success': True,
 'output': 'have amgm: 2 * 2 powr t * (6 * t) ≤ (2 powr t)⇧2 + (6 * t)⇧2',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. (2 powr t - 3 * t) * t / 4 powr t \\<le> 1 / 12'],
 'execution_time': 0.10209131240844727,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [41]:
await client.execute_command(session_id, 'have rearr: "12 * ((2 powr t - 3 * t) * t) = 2 * (2 powr t) * (6 * t) - (6 * t)^2"', timeout=120.0, lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. 12 * ((2 powr t - 3 * t) * t) = 2 * 2 powr t * (6 * t) - (6 * t)\\<^sup>2'],
 'execution_time': 0.10149717330932617,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [42]:
await client.execute_command(session_id, 'by (simp add: power2_eq_square algebra_simps)', timeout=120.0, lease_id=lease_id)

{'success': True,
 'output': 'have rearr:\n       12 * ((2 powr t - 3 * t) * t) = 2 * 2 powr t * (6 * t) - (6 * t)⇧2',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. (2 powr t - 3 * t) * t / 4 powr t \\<le> 1 / 12'],
 'execution_time': 0.1020505428314209,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [43]:
await client.execute_command(session_id, 'from amgm rearr hpow have key: "12 * ((2 powr t - 3 * t) * t) ≤ 4 powr t"', timeout=120.0, lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. 12 * ((2 powr t - 3 * t) * t) \\<le> 4 powr t'],
 'execution_time': 0.10160684585571289,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [44]:
await client.execute_command(session_id, 'by linarith', timeout=120.0, lease_id=lease_id)

{'success': True,
 'output': 'have key: 12 * ((2 powr t - 3 * t) * t) ≤ 4 powr t',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. (2 powr t - 3 * t) * t / 4 powr t \\<le> 1 / 12'],
 'execution_time': 0.10194897651672363,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [45]:
await client.execute_command(session_id, 'from key pos show ?thesis', timeout=120.0, lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. (2 powr t - 3 * t) * t / 4 powr t \\<le> 1 / 12'],
 'execution_time': 0.10175180435180664,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [46]:
await client.execute_command(session_id, 'by (simp add: field_simps)', timeout=120.0, lease_id=lease_id)

{'success': True,
 'output': 'show (2 powr t - 3 * t) * t / 4 powr t ≤ 1 / 12\nSuccessful attempt to solve goal by exported rule:\n  (2 powr t - 3 * t) * t / 4 powr t ≤ 1 / 12',
 'error': None,
 'subgoal_error': None,
 'subgoals': [],
 'execution_time': 0.10207438468933105,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [47]:
await client.execute_command(session_id, 'qed', timeout=120.0, lease_id=lease_id)

{'success': True,
 'output': 'theorem amc12b_2020_p22:\n (2 powr ?t - 3 * ?t) * ?t / 4 powr ?t ≤ 1 / 12',
 'error': None,
 'subgoal_error': None,
 'subgoals': [],
 'execution_time': 0.1020059585571289,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [48]:
await client.execute_command(session_id, 'end', timeout=120.0, lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': [],
 'execution_time': 0.10172724723815918,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

## Test session parallelism (`parallel_proofs` / `threads`)

Sessions are now created with configurable PIDE parallelism (`repl/src/main/scala/repl/session_manager.scala`):

- `ISABELLE_PARALLEL_PROOFS` (default `2`) — `2` forks nested `have`/`show` proof bodies so **independent `have`s are checked concurrently**; `1` = top-level proofs only; `0` = fully sequential.
- `ISABELLE_SESSION_THREADS` (default `4`) — caps the ML worker threads per session (deliberately not `0` = auto, so concurrent sessions don't oversubscribe cores and OOM the gateway).

Both are read from the **server's** environment at session creation — set them in `.env` at the repo root and restart the server to change them. Pooled sessions keep the options they were created with.

**The experiment:** one lemma with `N_HAVES` independent `have` steps, each proof sleeping `SLEEP_S` seconds inside an ML tactic before closing the goal. Submitted as **one chunk**, PIDE forks the `have` proofs, so wall time ≈ `ceil(N_HAVES / threads) * SLEEP_S`. The same steps submitted **one command at a time** form the sequential baseline (~`N_HAVES * SLEEP_S`), because each small-step waits for the node to finish checking before returning.

In [4]:
session_payload = await client.acquire_session(["Complex_Main"], "HOL")
session_id = session_payload.get("session_id")
lease_id = session_payload.get("lease_id")
session_payload

{'session_id': '9f22572d-ef71-4c95-af4d-95c157abe754',
 'created_at': 1781128631.4464405,
 'theories': ['Complex_Main'],
 'status': 'active',
 'reused': False,
 'lease_id': '27b151213841'}

In [5]:
N_HAVES = 4
SLEEP_S = 2

def slow_have(i: int) -> str:
    # Independent goal; the ML tactic sleeps, then closes it with simp.
    # With parallel_proofs=2 these `by` bodies are forked and run concurrently.
    return (
        f'have h{i}: "{i} + 0 = ({i}::nat)"\n'
        f"  by (tactic ‹(OS.Process.sleep (Time.fromSeconds {SLEEP_S}); simp_tac @{{context}} 1)›)"
    )

parallel_chunk = "\n".join(
    ["lemma par_demo: True", "proof -"]
    + [slow_have(i) for i in range(1, N_HAVES + 1)]
    + ["show True by simp", "qed"]
)
print(parallel_chunk)

lemma par_demo: True
proof -
have h1: "1 + 0 = (1::nat)"
  by (tactic ‹(OS.Process.sleep (Time.fromSeconds 2); simp_tac @{context} 1)›)
have h2: "2 + 0 = (2::nat)"
  by (tactic ‹(OS.Process.sleep (Time.fromSeconds 2); simp_tac @{context} 1)›)
have h3: "3 + 0 = (3::nat)"
  by (tactic ‹(OS.Process.sleep (Time.fromSeconds 2); simp_tac @{context} 1)›)
have h4: "4 + 0 = (4::nat)"
  by (tactic ‹(OS.Process.sleep (Time.fromSeconds 2); simp_tac @{context} 1)›)
show True by simp
qed


In [6]:
await client.enter_theory(session_id, "Par_Demo", lease_id=lease_id)
await client.execute_command(session_id, "theory Par_Demo imports Complex_Main begin", lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': [],
 'execution_time': 0.3339200019836426,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [7]:
import time

t0 = time.perf_counter()
res = await client.execute_command(session_id, parallel_chunk, timeout=120.0, lease_id=lease_id)
parallel_wall = time.perf_counter() - t0
print(f"one chunk, {N_HAVES} parallel haves: {parallel_wall:.1f}s "
      f"(sequential would be ~{N_HAVES * SLEEP_S}s of sleep alone)")
res

one chunk, 4 parallel haves: 6.9s (sequential would be ~8s of sleep alone)


{'success': True,
 'output': 'have h1: 1 + 0 = 1\nhave h2: 2 + 0 = 2\nhave h3: 3 + 0 = 3\nhave h4: 4 + 0 = 4\nshow True\nSuccessful attempt to solve goal by exported rule:\n  True\ntheorem par_demo: True',
 'error': None,
 'subgoal_error': None,
 'subgoals': [],
 'execution_time': 6.663745641708374,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [8]:
# Sequential baseline: same steps, one command per request.
# Each step waits for checking to finish, so the sleeps cannot overlap.
t0 = time.perf_counter()
await client.execute_command(session_id, "lemma seq_demo: True", timeout=120.0, lease_id=lease_id)
await client.execute_command(session_id, "proof -", timeout=120.0, lease_id=lease_id)
for i in range(1, N_HAVES + 1):
    await client.execute_command(session_id, slow_have(i), timeout=120.0, lease_id=lease_id)
await client.execute_command(session_id, "show True by simp", timeout=120.0, lease_id=lease_id)
await client.execute_command(session_id, "qed", timeout=120.0, lease_id=lease_id)
sequential_wall = time.perf_counter() - t0
print(f"step-by-step: {sequential_wall:.1f}s | one chunk: {parallel_wall:.1f}s "
      f"| speedup x{sequential_wall / parallel_wall:.1f}")

step-by-step: 26.6s | one chunk: 6.9s | speedup x3.8


In [9]:
await client.execute_command(session_id, "end", lease_id=lease_id)
await client.release_session(session_id, lease_id=lease_id)

{'success': True, 'session_id': '9f22572d-ef71-4c95-af4d-95c157abe754'}

To see the knobs bite, edit `.env`, restart the server, and rerun the chunk cell (the pool is rebuilt on restart, so new sessions pick up the options):

| `.env` setting | expected chunk wall time |
|---|---|
| `ISABELLE_PARALLEL_PROOFS=2`, `ISABELLE_SESSION_THREADS=4` | ≈ `SLEEP_S` (all 4 haves in flight) |
| `ISABELLE_PARALLEL_PROOFS=2`, `ISABELLE_SESSION_THREADS=2` | ≈ `2 × SLEEP_S` (two at a time) |
| `ISABELLE_PARALLEL_PROOFS=0` | ≈ `N_HAVES × SLEEP_S` (no forking — matches the step-by-step baseline) |

### Recorded results (2026-06-10)

Environment: rebuilt `isabellegym-isabelle-gym` image (Isabelle2025-2, linux/amd64 under qemu on Apple Silicon, `mem_limit: 12g`), `N_HAVES = 4`, `SLEEP_S = 2`. The first two rows are the cell outputs above; the third row reran the identical chunk after restarting the server with `ISABELLE_PARALLEL_PROOFS=0` (script: `claude-work/run-parallelism-demo/run_noparallel.py`), then the server was restarted again with the default parallel settings.

| server config | run | wall time | server `execution_time` |
|---|---|---|---|
| `parallel_proofs=2`, `threads=4` | one chunk, 4 parallel `have`s | **6.9 s** | 6.66 s |
| `parallel_proofs=2`, `threads=4` | step-by-step (one command per request) | **26.6 s** | — |
| `parallel_proofs=0` (forking off) | one chunk, same 4 `have`s | **24.9 s** | 24.72 s |

Takeaways:

- With forking on, the chunk runs **×3.8** faster than the step-by-step baseline (26.6 s → 6.9 s) — the speedup matches `N_HAVES = 4`, i.e. all four `have` bodies were checked concurrently.
- With `parallel_proofs=0` the *same chunk* degrades to 24.9 s ≈ `N_HAVES × SLEEP_S` + overhead — indistinguishable from the step-by-step baseline, confirming the speedup comes from PIDE proof forking and not from chunked submission itself.
- The parallel chunk does not hit the theoretical ~`SLEEP_S` floor (2 s) because the amd64 image runs under qemu emulation on this host, which inflates per-step checking overhead (visible also in the 26.6 s baseline vs ~8 s of pure sleep).

## Big-chunk verification with `verify_chunk`

The sledgehammer section above advanced the `amc12b_2020_p22` proof **one command per call** — many round-trips, and any single line could time out on its own with no indication of *which* line hung.

`verify_chunk` submits the **whole proof body in one call**. It is checked with per-session parallelism on, bounded by a **single** wall budget, and returns a **per-command status report** in source order — each command is `ok`, `failed`, `running`, or `unprocessed`. On a timeout the report is *partial* and names the still-`running` line (the likely loop) instead of surfacing an opaque timeout.

Below we verify the **exact same proof** from the sledgehammer section as a single chunk.

In [ ]:
# verify_chunk needs the theory already begun, so acquire a session, enter the
# theory, and process the header as a normal step first.
chunk_payload = await client.acquire_session(["Complex_Main"], "HOL")
chunk_sid = chunk_payload["session_id"]
chunk_lease = chunk_payload["lease_id"]
await client.enter_theory(chunk_sid, "amc12b_2020_p22", lease_id=chunk_lease)
await client.execute_command(
    chunk_sid,
    'theory amc12b_2020_p22 imports Complex_Main "HOL-Library.Code_Target_Numeral" begin',
    lease_id=chunk_lease,
)

In [ ]:
# The same theorem..qed proof from the sledgehammer section, as ONE chunk.
proof_chunk = r"""theorem amc12b_2020_p22: fixes t :: real shows "((2 powr t - 3 * t) * t) / (4 powr t) ≤ 1 / 12"
proof -
have hpow: "(2 powr t)^2 = 4 powr t"
by (metis powr_mult numeral_Bit0_eq_double power2_eq_square)
have pos: "(0::real) < 4 powr t" by simp
have amgm: "2 * (2 powr t) * (6 * t) ≤ (2 powr t)^2 + (6 * t)^2"
by (rule sum_squares_bound)
have rearr: "12 * ((2 powr t - 3 * t) * t) = 2 * (2 powr t) * (6 * t) - (6 * t)^2"
by (simp add: power2_eq_square algebra_simps)
from amgm rearr hpow have key: "12 * ((2 powr t - 3 * t) * t) ≤ 4 powr t"
by linarith
from key pos show ?thesis
by (simp add: field_simps)
qed
"""

# Single call, single wall budget; per-command status comes back in source order.
report = await client.verify_chunk(chunk_sid, proof_chunk, timeout=180.0, lease_id=chunk_lease)

# Built-in pretty printer in the async client — no custom loop needed.
client.print_chunk_report(report)
report

In [ ]:
# Finish the theory and return the session to the pool.
await client.execute_command(chunk_sid, 'end', timeout=120.0, lease_id=chunk_lease)
await client.close_session(chunk_sid, lease_id=chunk_lease)